### PR TITLE
WooCommerce: Add product search UI

### LINKED ISSUE #13655: WooCommerce: Add/Edit Product: Products List
Adds the list view for displaying a store's products. 

For v1, we are not offering quick or bulk edit. This should be a simple listing of a store's products, with search.

### PR DESCRIPTION
**Pull request by justinshreve**
This PR adds the UI for searching a site's products in the product list, including pagination, etc. It adds a search box above the list table.

screen shot 2017-06-20 at 9 36 24 am

To test:
* Run `npm test` and make sure all tests pass.
* Go to `http://calypso.localhost:3000/store/products/:site`
* Type a search term in the box and watch the list table update with the new results.
* Paginate your results and make sure that works.
* Type a search term that will return no results and make sure you get a message saying so.
* Hit the "x" in the search box (or clear your query) and see the normal list results return.

cc @kellychoffman @jameskoster since this PR contains some UI.

Closes #13655.

### CODE REVIEW

File: client/extensions/woocommerce/app/products/index.js [hunk 1]
```diff
@@ -101,11 +145,33 @@ class Products extends Component {
 		/>;
 	}
 
+	renderNoSearchResults() {
+		const { translate } = this.props;
+		return (
+			<div>
+				<p>
+					{ translate( 'No products match your search for {{searchTerm/}}.', {
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/index.js]: Hi! I've found a possible matching string that has already been translated **22** times:
`translate( 'No plugins match your search for {{searchTerm/}}.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-product-search-ui)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by kellychoffman** [comment]
Can we add a 'card' class to the search component? I tried, but failed.

**Comment by kellychoffman** [comment]
Actually, its a completely different component. `search-card` https://wpcalypso.wordpress.com/devdocs/design/search

**Comment by justinshreve** [comment]
@kellychoffman I have switched from `Search` to `SearchCard`.

File: client/extensions/woocommerce/app/products/index.js [hunk 2]
```diff
@@ -43,6 +57,11 @@ class Products extends Component {
 	// Total number of results per page (the API returns 10)
 	perPage = 10;
 
+	state = {
+		isSearching: false,
```
Comment (allendav) [inline client/extensions/woocommerce/app/products/index.js]: It seems like is less `isSearching` in the sense that a query is running that we are waiting for and more like `isShowingSearchResults` - is that right? Would you consider renaming?

**Comment by allendav** [comment]
There is a lot of `isSearching ? DO THIS : DO THAT` in this code. Would it simplify things by treating "not searching" as searching with a trivial query instead of having to manage two "modes"?

File: client/extensions/woocommerce/app/products/index.js [hunk 3]
```diff
@@ -55,17 +74,42 @@ class Products extends Component {
...
+	onSearch = ( query ) => {
+		const { site } = this.props;
+
+		if ( trim( query ) === '' ) {
+			this.setState( { isSearching: false, query: '' } );
+			this.props.clearProductSearch( site.ID );
+			return;
+		}
+
+		this.setState( { isSearching: true, query } );
```
Comment (allendav) [inline client/extensions/woocommerce/app/products/index.js]: Seems like isSearching is always true at the same time query is non trivial. Maybe you only need the query in state and could Boolean( trim( query ) ) elsewhere.  See also my suggestion however on treating "not searching" as a special case of searching with a "trivial" query.

File: client/extensions/woocommerce/app/products/products-list-search-results.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	requestedPage,
+	requestedPageLoaded,
+	query,
+	onSwitchPage,
+} ) => {
+	if ( currentPageLoaded === true && totalProducts === 0 ) {
+		return (
+			<div>
+				<p>
+					{ translate( 'No products match your search for {{searchTerm/}}.', {
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/products/products-list-search-results.js]: Hi! I've found a possible matching string that has already been translated **22** times:
`translate( 'No plugins match your search for {{searchTerm/}}.' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/woocommerce-product-search-ui)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by justinshreve** [comment]
@allendav I think I want to leave search as a separate state tree/request, as is, for now. One user benefit I like immediately being taken back to your place in the product list when closing out search. The other reason is I want to avoid changing out the list state code changes right now, in our last few days working.

However - I hear your concerns so I've refactored this PR quite a bit and the list UI code a bit to avoid all those checks and I like the way things are looking now -- feels much cleaner then before. Would you mind taking a look at what I'm doing now?

File: client/extensions/woocommerce/app/products/index.js [hunk 5]
```diff
@@ -6,42 +6,35 @@ import { bindActionCreators } from 'redux';
...
-		totalProducts: PropTypes.number,
-		fetchProducts: PropTypes.func,
+		fetchProducts: PropTypes.func.isRequired,
+		fetchProductSearchResults: PropTypes.func.isRequired,
+		clearProductSearch: PropTypes.func.isRequired,
 	};
 
-	// Total number of results per page (the API returns 10)
-	perPage = 10;
+	state = {
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/index.js]: should our components be stateful?  Could this be pushed to redux?  Not needed for now but something to think about post launch

File: client/extensions/woocommerce/app/products/index.js [hunk 6]
```diff
@@ -55,75 +48,41 @@ class Products extends Component {
...
-		);
+		this.setState( { query } );
+		this.props.fetchProductSearchResults( site.ID, 1, query );
 	}
 
 	render() {
 		const { className, site, translate } = this.props;
 		const classes = classNames( 'products__list', className );
+
+		const productsDisplay = trim( this.state.query ) === ''
```
Comment (belcherj) [inline client/extensions/woocommerce/app/products/index.js]: This multiline ternary feels icky.  I'm not saying don't do it but... :)

**Pull request merged**